File: OneCodeTeam/How-to-use-Windows-Azure-Service-Management-27261/[VB]-How to use Windows Azure Service Management Libraries to create Azure VM/C#/CSAzureCreateVMUseManagementLibraries/MyVirtualMachineExtension.cs
/****************************** Module Header ******************************\
* Module Name: MyVirtualMachineExtension.cs
* Project:     CSAzureCreateVMUseMangementLibaries
* Copyright (c) Microsoft Corporation.
* 
* We use this Enxtension class make the method similar to the REST API
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/en-us/openness/licenses.aspx#MPL
* All other rights reserved.
* 
* THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
* EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
* WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\***************************************************************************/


using Microsoft.WindowsAzure.Management.Compute;
using Microsoft.WindowsAzure.Management.Compute.Models;
using Microsoft.WindowsAzure;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;

namespace CSAzureCreateVMUseMangementLibaries
{
    public static class MyVirtualMachineExtension
    {
        /// <summary>
        /// Instantiate a new VM Role
        /// </summary>
        public static Role CreateVMRole(this IVirtualMachineOperations client, string cloudServiceName, string roleName, VirtualMachineRoleSize roleSize, string userName, string password, OSVirtualHardDisk osVHD)
        {
            Role vmRole = new Role
            {
                RoleType = VirtualMachineRoleType.PersistentVMRole.ToString(),
                RoleName = roleName,
                Label = roleName,
                RoleSize = roleSize,
                ConfigurationSets = new List<ConfigurationSet>(),
                OSVirtualHardDisk = osVHD
            };
            ConfigurationSet configSet = new ConfigurationSet
            {
                ConfigurationSetType = ConfigurationSetTypes.WindowsProvisioningConfiguration,
                EnableAutomaticUpdates = true,
                ResetPasswordOnFirstLogon = false,
                ComputerName = roleName,
                AdminUserName = userName,
                AdminPassword = password,
                InputEndpoints = new BindingList<InputEndpoint>
                {
                    new InputEndpoint { LocalPort = 3389, Name = "RDP", Protocol = "tcp" },
                    new InputEndpoint { LocalPort = 80, Port = 80, Name = "web", Protocol = "tcp" }
                }
            };
            vmRole.ConfigurationSets.Add(configSet);
            return vmRole;
        }

        public static OSVirtualHardDisk CreateOSVHD(this IVirtualMachineImageOperations operation, string cloudserviceName, string vmName, string storageAccount, string imageFamiliyName)
        {
            try
            {
                var osVHD = new OSVirtualHardDisk
                {
                    MediaLink = GetVhdUri(string.Format("{0}.blob.core.windows.net/vhds", storageAccount), cloudserviceName, vmName),
                    SourceImageName = GetSourceImageNameByFamliyName(operation, imageFamiliyName)
                };
                return osVHD;
            }
            catch (CloudException e)
            {

                throw;
            }

        }

        private static string GetSourceImageNameByFamliyName(this IVirtualMachineImageOperations operation, string imageFamliyName)
        {
            var disk = operation.List().Where(o => o.ImageFamily == imageFamliyName).FirstOrDefault();
            if (disk != null)
            {
                return disk.Name;
            }
            else
            {
                throw new CloudException(string.Format("Can't find {0} OS image in current subscription"));
            }
        }

        private static Uri GetVhdUri(string blobcontainerAddress, string cloudServiceName, string vmName, bool cacheDisk = false, bool https = false)
        {
            var now = DateTime.UtcNow;
            string dateString = now.Year + "-" + now.Month + "-" + now.Day;

            var address = string.Format("http{0}://{1}/{2}-{3}-{4}-{5}-650.vhd", https ? "s" : string.Empty, blobcontainerAddress, cloudServiceName, vmName, cacheDisk ? "-CacheDisk" : string.Empty, dateString);
            return new Uri(address);
        }

        public static void CreateVMDeployment(this IVirtualMachineOperations operations, string cloudServiceName, string deploymentName, List<Role> roleList, DeploymentSlot slot = DeploymentSlot.Production)
        {

            try
            {
                VirtualMachineCreateDeploymentParameters createDeploymentParams = new VirtualMachineCreateDeploymentParameters
                {

                    Name = deploymentName,
                    Label = cloudServiceName,
                    Roles = roleList,
                    DeploymentSlot = slot
                };
                operations.CreateDeployment(cloudServiceName, createDeploymentParams);
            }
            catch (CloudException e)
            {
                throw;
            }
        }
    }
}

File: OneCodeTeam/How-to-use-Windows-Azure-Service-Management-27261/[VB]-How to use Windows Azure Service Management Libraries to create Azure VM/VB/VBAzureCreateVMUseManagementLibraries/MyVirtualMachineExtension.vb
'***************************** Module Header ******************************\
' Module Name: MyVirtualMachineExtension.vb
' Project:     VBAzureCreateVMUseMangementLibraries
' Copyright (c) Microsoft Corporation.
' 
' We use this Enxtension class to make these methods similar to the REST API.
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/en-us/openness/licenses.aspx#MPL
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'**************************************************************************/

Imports Microsoft.WindowsAzure
Imports Microsoft.WindowsAzure.Management.Compute
Imports Microsoft.WindowsAzure.Management.Compute.Models
Imports System.ComponentModel
Imports System.Runtime.CompilerServices

Module MyVirtualMachineExtension
    ''' <summary>
    ''' Instantiation a new VM Role
    ''' </summary>
    <Extension()>
    Public Function CreateVMRole(client As IVirtualMachineOperations, cloudServiceName As String, roleName As String, roleSize As VirtualMachineRoleSize, userName As String, password As String, _
        osVHD As OSVirtualHardDisk) As Role
        Dim vmRole As New Role() With { _
             .RoleType = VirtualMachineRoleType.PersistentVMRole.ToString(), _
             .RoleName = roleName, _
             .Label = roleName, _
             .RoleSize = roleSize, _
             .ConfigurationSets = New List(Of ConfigurationSet)(), _
             .OSVirtualHardDisk = osVHD _
        }
        Dim configSet As New ConfigurationSet() With { _
             .ConfigurationSetType = ConfigurationSetTypes.WindowsProvisioningConfiguration, _
             .EnableAutomaticUpdates = True, _
             .ResetPasswordOnFirstLogon = False, _
             .ComputerName = roleName, _
             .AdminUserName = userName, _
             .AdminPassword = password, _
             .InputEndpoints = New BindingList(Of InputEndpoint)() From { _
                New InputEndpoint() With { _
                     .LocalPort = 3389, _
                     .Name = "RDP", _
                     .Protocol = "tcp" _
                }, _
                New InputEndpoint() With { _
                     .LocalPort = 80, _
                     .Port = 80, _
                     .Name = "web", _
                     .Protocol = "tcp" _
                } _
            } _
        }
        vmRole.ConfigurationSets.Add(configSet)
        Return vmRole
    End Function

    <Extension()>
    Public Function CreateOSVHD(operation As IVirtualMachineImageOperations, cloudserviceName As String, vmName As String, storageAccount As String, imageFamiliyName As String) As OSVirtualHardDisk
        Try
            Dim osVHD = New OSVirtualHardDisk() With { _
                 .MediaLink = GetVhdUri(String.Format("{0}.blob.core.windows.net/vhds", storageAccount), cloudserviceName, vmName), _
                 .SourceImageName = GetSourceImageNameByFamliyName(operation, imageFamiliyName) _
            }
            Return osVHD
        Catch e As CloudException

            Throw
        End Try

    End Function

    <Extension()>
    Private Function GetSourceImageNameByFamliyName(operation As IVirtualMachineImageOperations, imageFamliyName As String) As String
        Dim disk = operation.List().Where(Function(o) o.ImageFamily = imageFamliyName).FirstOrDefault()
        If disk IsNot Nothing Then
            Return disk.Name
        Else
            Throw New CloudException(String.Format("Can't find {0} OS image in current subscription"))
        End If
    End Function

    <Extension()>
    Private Function GetVhdUri(blobcontainerAddress As String, cloudServiceName As String, vmName As String, Optional cacheDisk As Boolean = False, Optional https As Boolean = False) As Uri
        Dim now = DateTime.UtcNow
        Dim dateString As String = Convert.ToString(now.Year) & "-" & Convert.ToString(now.Month) & "-" & Convert.ToString(now.Day)

        Dim address = String.Format("http{0}://{1}/{2}-{3}-{4}-{5}-650.vhd", If(https, "s", String.Empty), blobcontainerAddress, cloudServiceName, vmName, If(cacheDisk, "-CacheDisk", String.Empty), _
            dateString)
        Return New Uri(address)
    End Function

    <Extension()>
    Public Sub CreateVMDeployment(operations As IVirtualMachineOperations, cloudServiceName As String, deploymentName As String, roleList As List(Of Role), Optional slot As DeploymentSlot = DeploymentSlot.Production)
        Try
            Dim createDeploymentParams As New VirtualMachineCreateDeploymentParameters() With { _
                .Name = deploymentName, _
                 .Label = cloudServiceName, _
                 .Roles = roleList, _
                 .DeploymentSlot = slot _
            }
            operations.CreateDeployment(cloudServiceName, createDeploymentParams)
        Catch e As CloudException
            Throw
        End Try
    End Sub
End Module
